README.md
Inital information compiled from lots of good stuff around Scripta at litecointalk
[fourm posts](https://litecointalk.org/index.php?topic=9908.msg143787#msg143787)

1. Started with Scripta 1.1 [image](http://www.lateralfactory.com/download.php?file=scripta-1_1.tgz)
    * source files in image do not match current Scripta [repo](https://github.com/scriptamining/scripta.git)  
    * rsync github files to image to get starting code base (all uses of "minepeon" replaced with "scripta")
    * origianl MinePeon [project](http://minepeon.com/index.php/Main_Page) has lots of good PI setup and helpful info
    
2. Update raspberry to newest kernel (3.10.25+) and added 'slub_debug=FP' to bootline.conf
    * this should fix the wierd USB debug error logging issues.   
    * TODO: TLS Warning (BRANCH=next rpi-update)  

3. Set ssh port to 22.  Should probably turn off root ssh access and change password.  
    ssh root@10.0.1.28  
    password: scripta  
    
4. Change locales to en_US

5. Add wifi support (wlan0)
    * set ssid and psk in 'network' block at '/etc/wpa_supplicant/wpa_supplicant.conf'

6. Built GridSeed GS3355 specific version of cgminer with mulit-frequency support from [repo](https://github.com/girnyau/cgminer-gc3355)
    * modified cgminer-gc3355 to report Frequency and Serial in API calls: [repo](https://github.com/mox235/cgminer-gc3355)

7. Edit '/opt/scripta/startup/miner-start.sh' to use cgminer-gc3355

8. Remove Scripta LTC donation address.  Disable MinePeon BTC hash time donation option.  Need to determine the right thing to do since _most_ of the web interface stuff is directly from the [MinePeonUI project](https://github.com/MineForeman/zArchive-MinePeonWebUI.git)
    * other Scripta rebranded PI images at [Crypto Pros](http://www.cryptopros.com/2014/03/gridseed-dual-miner-first-look-amazing.html) 
    * and [Hash Master](https://hash-master.com/blog/using-your-raspberry-pi-as-a-gridseed-mining-controller/)
    * different PI web-based controller for Gridseed at [Hashra](https://github.com/HASHRA)

9. Fix pool URL JSON encoding.  Add back miner config name/values settings from MinePeonUI.  All cgminer settings can be changed or added from miner form. 
    * ability to set per GSD frequency based on Serial # from list: 
    ```
        static const int opt_frequency[] = {
            700,  706,  713,  719,  725,  731,  738,  744,
            750,  756,  763,  769,  775,  781,  788,  794,
            800,  813,  825,  838,  850,  863,  875,  888,
            900,  913,  925,  938,  950,  963,  975,  988,
           1000, 1013, 1025, 1038, 1050, 1063, 1075, 1088,
           1100, 1113, 1125, 1138, 1150, 1163, 1175, 1188,
           1200, 1213, 1225, 1238, 1250, 1263, 1275, 1288,
           1300, 1313, 1325, 1338, 1350, 1363, 1375, 1388,
           1400,
             -1
        };
    ```
10. Modify Status table
    * show KHs instead of MHs
    * replace Name with Serial Number
    * replace Temperature with Frequency
    
11. Open Issues
    * miner commands menu seem broken   
    * DAcc seems crazy
    * maybe reported [hashrate](http://cryptomining-blog.com/1760-what-is-the-actual-hashrate-you-get-from-your-gridseed-asic/) is not quite accurate



changes.md
Inital information compiled from lots of good stuff around Scripta at litecointalk
[fourm posts](https://litecointalk.org/index.php?topic=9908.msg143787#msg143787)

1. Started with Scripta 1.1 [image](http://www.lateralfactory.com/download.php?file=scripta-1_1.tgz)
    source file in image do not match [repo]([https://github.com/scriptamining/scripta.git])
    rsync git files to image to get starting code base (all uses of "minepeon" replaced with "scripta" - why?)
    
2. Update raspberry to newest kernel (3.10.25+) and added 'slub_debug=FP' to bootline.conf
    This should maybe fix the wierd USB debug error logging issues. 
    TODO: TLS Warning (BRANCH=next rpi-update)

3. Set ssh port to 22.  Should probably turn off root ssh access and change password.
    ssh root@10.0.1.28
    password: scripta
    
4. Change locales to en_US

5. Add wifi support (wlan0)
    set ssid and psk in 'network' block at '/etc/wpa_supplicant/wpa_supplicant.conf'

6. Built GridSeed GS3355 specific version of cgminer with mulit-frequency support from [repo](https://github.com/girnyau/cgminer-gc3355)

7. Edit '/opt/scripta/startup/miner-start.sh' to use cgminer-gc3355

8. Remove Scripta LTC donation address.  Disable MinePeon BTC hash time donation option.  Need to determine the right thing to do since _all_ of the web interface stuff is directly from the [MinePeon project](http://minepeon.com/index.php/Main_Page).  

9. Issues:
    Miner commands seem broken.  
    Pool URL is sometimes wrong.  
    Replace Name with Serial number.
    Replace Freq with Temp.
    Khash instead of Mhash.
    RRD plots not updaing.  
    Add option to choose miner exe.
    Clear HW error after some time (or accept?).

10. Add GridSeed specific options to Miner form.  

{
"pools" : [
	{	
		"name" : "WeMineLTC",
		"url" : "stratum+tcp://usa.wemineltc.com:3336",
		"user" : "mox235.1",
		"pass" : "1"
	},
	{
		"name" : "WeMineLTC",
		"url" : "stratum+tcp://united.wemineltc.com:3335",
		"user" : "mox235.1",
		"pass" : "1"
	},
	{
		"name" : "ScryptGuild",
		"url" : "stratum+tcp://stratum.scryptguild.com:3333",
		"user" : "mox235_gridseed",
		"pass" : "1"
	}
],
"api-allow" : "W:127.0.0.1,192.168.1/24",
"api-listen" : true,
"failover-only" : true,
"scrypt" : true,
"gridseed-options" : "baud=115200,chips=5",
"gridseed-freq" : "6D8512995148=950,6D81116F5650=925,6D9021724857=925,6D82475A4857=900,6D973F855057=900,6D8B54835251=900"
}

var/www/f_status.php
<?php
session_start();

if ( !isset($_SESSION['_logged_']) || $_SESSION['_logged_'] === false ) {
	die();
} 

/*
f_status gets values that people want to see in realtime
returns success, status data and errors
*/
header('Content-type: application/json');

include('inc/cgminer.inc.php');
include('inc/ChromePhp.php');

// Miner data
//$r['summary'] = cgminer('summary', '')['SUMMARY'];
$devs=cgminer('devs');
$pools=cgminer('pools');

if(!empty($devs['data']['DEVS'])){
  $r['status']['devs'] = $devs['data']['DEVS'];
}
else{
  $r['status']['devs'] = array();
}
if(!empty($pools['data']['POOLS'])){
  $r['status']['pools'] = $pools['data']['POOLS'];
  $r['status']['minerUp'] = true;
  $r['status']['minerDown'] = false;
}
else{
  $r['status']['pools'] = array();
  $r['status']['minerUp'] = false;
  $r['status']['minerDown'] = true;
}

// Debug miner data
if(!empty($_REQUEST['dev']) && $r['status']['minerUp']){
  $r['status']['devs'][]=array('Name'=>'Hoeba','ID'=>0,'Temperature'=>rand(20,35),'MHS5s'=>rand(80000,100000),'MHSav'=>rand(90000,100000),'LongPoll'=>'N','Getworks'=>200,'Accepted'=>rand(70,200),'Rejected'=>rand(1,10),'HardwareErrors'=>rand(0,50),'Utility'=>1.2,'LastShareTime'=>time()-rand(0,10));
  $r['status']['devs'][]=array('Name'=>'Debug','ID'=>1,'Temperature'=>rand(20,35),'MHS5s'=>rand(40000,50000),'MHSav'=>rand(45000,50000),'LongPoll'=>'N','Getworks'=>1076,'Accepted'=>1324,'Rejected'=>1,'HardwareErrors'=>46,'Utility'=>1.2,'LastShareTime'=>time()-rand(0,40));
  $r['status']['devs'][]=array('Name'=>'Wut','ID'=>2,'Temperature'=>rand(20,35),'MHS5s'=>rand(6000,9000),'MHSav'=>rand(7000,8000),'LongPoll'=>'N','Getworks'=>1076,'Accepted'=>1324,'Rejected'=>1,'HardwareErrors'=>46,'Utility'=>1.2,'LastShareTime'=>time()-rand(0,300));
  $r['status']['devs'][]=array('Name'=>'Wut','ID'=>4,'Temperature'=>rand(20,35),'MHS5s'=>0,'MHSav'=>0,'LongPoll'=>'N','Getworks'=>400,'Accepted'=>0,'Rejected'=>0,'HardwareErrors'=>0,'Utility'=>0,'LastShareTime'=>time()-rand(400,900));
  $r['status']['devs'][]=array('Name'=>'More','ID'=>3,'Temperature'=>rand(20,35),'MHS5s'=>rand(500,1000),'MHSav'=>rand(600,800),'LongPoll'=>'N','Getworks'=>1076,'Accepted'=>1324,'Rejected'=>1,'HardwareErrors'=>46,'Utility'=>1.2,'LastShareTime'=>time()-rand(0,300));
  $r['status']['pools'][]=array('POOL'=>5,'URL'=>'http://stratum.mining.eligius.st:3334','Status'=>'Alive','Priority'=>9,'LongPoll'=>'N','Getworks'=>10760,'Accepted'=>50430,'Rejected'=>60,'Discarded'=>21510,'Stale'=>0,'GetFailures'=>0,'RemoteFailures'=>0,'User'=>'1BveW6ZoZmx31uaXTEKJo5H9CK318feKKY','LastShareTime'=>1375501281,'Diff1Shares'=>20306,'ProxyType'=>'','Proxy'=>'','DifficultyAccepted'=>20142,'DifficultyRejected'=>24,'DifficultyStale'=>0,'LastShareDifficulty'=>4,'HasStratum'=>true,'StratumActive'=>true,'StratumURL'=>'stratum.mining.eligius.st','HasGBT'=>false,'BestShare'=>40657);
}

$devices = 0;
$MHSav = 0;
$MHS5s = 0;
$Accepted = 0;
$Rejected = 0;
$HardwareErrors = 0;
$Utility = 0;

// ChromePhp::log($r['status']);

if(!empty($r['status']['devs'])){
  foreach ($r['status']['devs'] as $id => $dev) {
    $devices += $dev['MHS5s']>0?1:0; // Only count hashing devices
    $MHS5s += $dev['MHS5s'];
    $MHSav += $dev['MHSav'];
    $Accepted += $dev['Accepted'];
    $Rejected += $dev['Rejected'];
    $HardwareErrors += $dev['HardwareErrors'];
    $Utility += $dev['Utility'];
    $r['status']['devs'][$id]['TotalShares']=$dev['Accepted']+$dev['Rejected']+$dev['HardwareErrors'];
  }
}

$r['status']['dtot']=array(
  'devices'=>$devices,
  'MHS5s'=>$MHS5s,
  'MHSav'=>$MHSav,
  'Accepted'=>$Accepted,
  'Rejected'=>$Rejected,
  'HardwareErrors'=>$HardwareErrors,
  'Utility'=>$Utility,
  'TotalShares'=>$Accepted+$Rejected+$HardwareErrors);

// CPU intensive stuff
if(!empty($_REQUEST['all'])){
  $ret = sys_getloadavg();
  $r['status']['pi']['load'] = $ret[2];
  $ret = explode(' ', exec('cat /proc/uptime'));
  $r['status']['pi']['uptime'] = $ret[0];
  $r['status']['pi']['temp'] = exec('cat /sys/class/thermal/thermal_zone0/temp')/1000;

  // What other interesting stuff is in summary?
  $summary=cgminer('summary');
  if(!empty($summary['data']['SUMMARY'][0]['Elapsed'])){
    $r['status']['uptime'] = $summary['data']['SUMMARY'][0]['Elapsed'];
  }
  else{
    $r['status']['uptime'] = 0;
  }
}

$r['status']['time'] = time();

//ChromePhp::log($r);

echo json_encode($r);
?>
